### PR TITLE
Improvements on package

### PR DESCRIPTION
**Pull request by ElsLommelen**
I'm dealing with issues #34 and #30 to make some improvements on the package:
- more informative error messages
- automate the checks and the pkgdown website build
- deal with the issue that the vignette can only be build if the database is accessible

I have some items to discuss:
- The solution I have for the vignette at the moment, is to not evaluate (and not show the results) when the package is built without database access.  This also means no results be shown on the website, unless I search a solution for this.  I suppose we at least want to show the results of the code there?
- I noticed I initially added a MIT licence, I don't know anymore why.  Do we keep it, or do we change it to GPL3?
- `get_inboveg_recordings()` gets the lintr error of high cyclomatic complexity (too many use of `if`).  @ElsDeBie @hansvancalster Do you want to deal with it yourselves, or do you want me to have a try?  In this case I need a volunteer to check the function into detail afterwards... ;-)
- The checklist package advises to write the title (which is shown on the upper right of the main page on github) in title case (all verbs and nowns in capitals).  Personally I'm not in favour of this style... Should we do it anyway to meet the checklist requirements, or should we ignore this note?
- I think we definitely should rename the `master` branch to `main`, when do you want me to do it?  (Just let me know if you are working on the package at the moment or planning to do so in near future, otherwise I do it one of the next days.)

@ThierryO If you wish, you can already review #34.

### CODE REVIEW

**Comment by ElsDeBie** [comment]
Hi, I can help with the function 'get_inboveg_recordings' as I want to work on new functions for inboveg as well in this part of the inbodb package.

**Comment by ElsLommelen** [comment]
@ElsDeBie Great!
I've been thinking a bit on how to simplify it, just some ideas:
- instead of checking if arguments are missing, you could already give it an appropriate value in the declaration (as you did with collect and multiple), for instance `%` -> we don't need the first block of `if` statements anymore
- for the second block of `if` statements,
    - we could just add the `multiple` in the different `if` statements, avoiding to have different nested levels,
    - and/or we could integrate them in the `if` statements below the `common_part` query
    - and/or we could replace them by an `assert_that()` where the combination of the `if`'s is tested (or integrate the `if` statement in the existing `assert_that()`); in this case you may want to use the argument `msg =` to give a clear message to the user, because the automatically generated message is in these cases often not clear (just test if they are clear enough, otherwise provide your own)
- instead of writing separate code for every combination of variables, we could try pasting an additional part to the query for each variable that is missing, similar to [this example](https://github.com/inbo/LSVI/blob/develop/R/selecteerIndicatoren.R) (fortunately we don't need the `voegwoord` part here, as the `common_part` already contains WHERE, and yes, you may want to keep using `glue_sql`, the example is rather old).  (In this way you could even combine exact matches with partial matches, if it is appropriate for users to have this option.)
- the `load`-functions in [this package](https://github.com/inbo/forrescalc/tree/master/R), together with the function `query_database()` give another example of combining parts of a query.  I'm not sure it is suitable to be used in this situation, but maybe it inspires you anyway?

Do you want to continu working in this PR, or will you make another one later on?  In the latter case, I will mute the cyclomatic complexity error temporarily to be able to merge.  And than maybe it is a good idea to rename `master` to `main` before you start working on it, so you can change it locally beforehand.

**Comment by ElsDeBie** [comment]
I will work on a new branch ... but I must clean up my branches in Rstudio as I don't see any of them in the github?

**Comment by ElsLommelen** [comment]
Yes, see [here](https://inbo.github.io/git-course/workflow_review.html) below.  There should be 5 branches at the moment, and actually `fb-traits-bron` could be removed on GitHub.  And than I'll maybe first change master to main, I don't know yet what we have to do locally after changing this.

**Comment by hansvancalster** [comment]
@ElsLommelen here are some notes I made for renaming branches:

`git branch -m new-branch-name` Rename a local branch  while having checked out the branch to be renamed

Followed by either 

`git push -u origin -u new-branch-name` Push the renamed branch to remote and track it

or only set the tracking branch without pushing:

`git branch -u origin/new-branch-name` (equivalent with `git branch -u origin/new-branch-name new-branch-name` if you are working locally in `new-branch-name`)




`git push origin --delete <old_name>` Delete the <old_name> remote branch

Some general checks you can do after that:

`git fetch` Updates your remote tracking branches

`git branch -vv` Shows tracking and synchronization status of your local branches

`git pull` Updates current local branch if you're not yet up to date

**Comment by ElsLommelen** [comment]
Thanks @hansvancalster, I guess I'll just use [the suggestion given by checklist](https://github.com/github/renaming) to change the branch name on GitHub.  And then maybe try `git fetch` to fix it locally?  I'll test it and post the manual here.  ;-)  Just don't start working locally on the package now.

**Comment by ElsLommelen** [comment]
Ok, GitHub itself gave me the following commands to change it locally:
```
git branch -m master main
git fetch origin
git branch -u origin/main main
git remote set-head origin -a
```
You will probably get them by email as well.

**Comment by ElsLommelen** [comment]
Tried, and it works without any problems.  I would say: it takes 5 minutes, so don't hesitate to change it in your own repo's as well.  ;-)

File: R/connect_inbo_dbase.R [hunk 1]
```diff
@@ -58,13 +60,38 @@ connect_inbo_dbase <- function(database_name, autoconvert_utf8 = TRUE) {
...
+    tryCatch(
+        conn <- dbConnect(odbc(),
+                          driver = sql_driver,
+                          server = server,
+                          port = 1433,
+                          database = database_name,
+                          encoding = encoding,
+                          trusted_connection = "YES"),
+        error = function(e) {
+            if (grepl("connection to SQL Server", e)) {
```
Comment (ThierryO) [inline R/connect_inbo_dbase.R]: You could write this as `assert_that(!grepl("connection to SQL Server", e), msg = ...)` Which avoids the if statements.

```
assert_that() # connection
assert_that() # login
stop(e)
```

File: R/get_inboveg_recordings.R [hunk 2]
```diff
@@ -84,19 +87,16 @@ get_inboveg_recordings <- function(
   if (!multiple) {
     if (missing(survey_name)) {
       survey_name <- "%"
```
Comment (ThierryO) [inline R/get_inboveg_recordings.R]: set this as a default value in the function arguments. So you don't need the if

Comment (ElsLommelen) [reply R/get_inboveg_recordings.R]: That's what I already suggested [here](https://github.com/inbo/inbodb/pull/35#issuecomment-872830331).  ;-)

Comment (ElsLommelen) [reply R/get_inboveg_recordings.R]: Thanks anyway to have a look at it, in your other suggestion I had missed the simplification myself...

**Comment by ElsLommelen** [comment]
Regarding the vignette `get_data_inboveg` I added the following solution:
- I made separate chunks for outputs, where eval is always TRUE
- during run when database access, all chunks are evaluated and all variables are saved in a `.RData` object
- during run when no database access, the `.RData` object is loaded and only chunks with outputs are evaluated (and here the loaded variables are shown)

There is a special case in one chunk (with `relations_N2000meetnet`), where the object is a pointer to the database query (which  doesn't give an output if the database is not available).  Here I write the output of the object to another variable (which is also saved in the `.Rdata` file), and add this output when there is no database access.  Unfortunately this output has a larger font and a black font colour... To be solved next week.

Some advise to work with it:
- make sure to render the vignette with database access after you made changes in it (and you may want to do it regularly to also capture database changes) -> just knitting does the job, you don't have to do any manual actions
- when adding new code, look at the example chunks:
    - if you want the code to run while connected to the database (e.g. passing code to a variable that you want to show later), don't add any setting on the chunk (will be evaluated when database access)
    - if you want to show the content of a variable, use `eval=TRUE`
    - if you don't want to evaluate the code at all, use `eval=FALSE`
- to be sure everything is presented well, knit the vignette first with database access (and check it), and knit it again afterwards without database access (and check it again)

**Comment by ElsDeBie** [comment]
if I set all necessary parameters at %, than multiple must be TRUE as default ...which means that in the other functions of INBOVEG there is a difference (multiple = FALSE). Is this okay, or beter to make it more equal?

**Comment by ElsLommelen** [comment]
Eh, if multiple = FALSE (`!multiple` in code), you set all parameters to `%` in your code, so I don't understand why you have to change the default of multiple to TRUE?  When looking at your code, to me it seems that `%` would be a convenient default for multiple = FALSE.  (I'm not sure which would be a good default for multiple = TRUE, but instead of check for missing, it can be checked if it is equal to `%`, so there is certainly no problem to be expected when using it.)

**Comment by ElsLommelen** [comment]
@ThierryO  Could you please help me with this?  These checks fail during CI while `checklist::check_package()` locally only gives a note.
- check package: locally I don't get the error that suggests to run `devtools::document()`.  Running `devtools::document()` locally (with roxygen2 7.1.2 installed) rewrites the NAMESPACE file, but apparently without any changes, as the file does not appear in the git window (also not after refreshing)
- R-CMD-check-OS ubuntu: there seems to be a problem with the installation of rcmdcheck in ubuntu, so to be solved in checklist?

**Comment by ThierryO** [comment]
It is often easier to track this in the raw logs. You can find those when you click on the cog wheel at the top right of the formatted log. Searching for `rcmdcheck` in the Ubuntu 20.04 logs revealed

```
2022-01-03T15:25:56.5220453Z ERROR: dependency ‘curl’ is not available for package ‘rcmdcheck’
```
I suggest that you try to find out why curl is not available by looking at the raw logs. Then try to fix that. Feel free to ask more pointer in case you get stuck.

**Comment by ElsLommelen** [comment]
@ThierryO Thanks to your hint I succeeded in solving the problem for Ubuntu 20.4, but not for Ubuntu 18.4, and the check package problem stays a real mystery for me: what does the namespace test check?  How do I have to update the namespace while `devtools::document()` doesn't generate a new version?  To me that's a drawback of checklist: it clearly says what to do, but if doing this doesn't work, as a user I don't have any clue how I can solve the problem because I even don't know what is tested (even not when sifting the logging into detail).

**Comment by ElsLommelen** [comment]
@ThierryO I finally solved the problem in the check package action, but I don't find a solution for the problem in R-CMD-check-OS of ubuntu-18.04.  Could it be that some remnants of the R-CMD-check-OS of ubuntu-20.04 cause the problem?  And how could I solve this?  Changing the order of the OS under strategy/matrix/config apparently doesn't change the order of performance.  Or is something else causing the problem?  Any help is appreciated.  (I hope to be able to finish this PR asap, as the package will be used in the coding club next week...)

**Comment by ThierryO** [comment]
Ubuntu 18.04 has reached [end of life](https://9to5linux.com/ubuntu-18-04-flavors-reach-end-of-life-users-urged-to-upgrade-to-20-04-lts). [Ubuntu 22.04](https://linuxconfig.org/ubuntu-22-04-features-and-release-date) will release in a few month.

Therefore I'd suggest to test oldrel on Ubuntu 20.04 instead. I'll update this in inbo/checklist#76.

File: .github/workflows/check_on_different_r_os.yml [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+    name: ${{ matrix.config.os }} (${{ matrix.config.r }})
+
+    strategy:
+      fail-fast: false
+      matrix:
+        config:
+          - {os: macOS-latest,   r: 'release'}
+          - {os: windows-latest, r: 'release'}
+          - {os: ubuntu-20.04,   r: 'oldrel',  rspm: "https://packagemanager.rstudio.com/cran/__linux__/xenial/latest"}
```
Comment (ThierryO) [inline .github/workflows/check_on_different_r_os.yml]:
```suggestion
          - {os: ubuntu-20.04,   r: 'oldrel',  rspm: "https://packagemanager.rstudio.com/cran/__linux__/focal/latest"}
```

File: R/connect_inbo_dbase.R [hunk 4]
```diff
@@ -6,11 +6,13 @@
...
 #'
 #' For more information, refer to
-#' \href{https://inbo.github.io/tutorials/tutorials/r_database_access/}{this tutorial}.
+#' \href{https://inbo.github.io/tutorials/tutorials/r_database_access/}{this
+#' tutorial}.
 #'
 #' @param database_name char Name of the INBO database you want to connect
 #' @param autoconvert_utf8 Should the encoding of the tables that are retrieved
-#' from the database be adapted to ensure correct presentation?  Defaults to TRUE.
+#' from the database be adapted to ensure correct presentation?  Defaults to
```
Comment (ThierryO) [inline R/connect_inbo_dbase.R]: I prefer to start every sentence on a new line. Then changing one sentence will not influence other sentences as they are on different line. No need to change this now. I tend to do this only when I update documentation.

File: R/connect_inbo_dbase.R [hunk 5]
```diff
@@ -136,7 +162,7 @@ on_connection_opened <- function(connection, code, type) {
         },
 
         # table preview code
-        previewObject = function(rowLimit, ...) {
+        previewObject = function(rowLimit, ...) { #nolint
```
Comment (ThierryO) [inline R/connect_inbo_dbase.R]: Only ignore the `object_name_linter()`
```suggestion
        previewObject <- function(rowLimit, ...) { #nolint: object_name_linter.
```

File: vignettes/get_data_inboveg.Rmd [hunk 6]
```diff
@@ -125,34 +148,37 @@ all_header_info <- get_inboveg_header(con)
 
 
 ## Recording information
-The function  `get_inboveg_recordings` queries the INBOVEG database for relevé information (which species were recorded in which plots and in which vegetation layers with which cover) for one or more surveys.
+The function  `get_inboveg_recording` queries the INBOVEG database for relevé information (which species were recorded in which plots and in which vegetation layers with which cover) for one or more surveys.
```
Comment (ThierryO) [inline vignettes/get_data_inboveg.Rmd]: Add brackets to make it look like a function. Then pkgdown can detect it and automatically create a link to the helppage.
```suggestion
The function  `get_inboveg_recording()` queries the INBOVEG database for relevé information (which species were recorded in which plots and in which vegetation layers with which cover) for one or more surveys.
```

File: vignettes/get_data_inboveg.Rmd [hunk 7]
```diff
@@ -222,119 +251,132 @@ all_codes <- get_inboveg_classification(con)
 
 
 ## Qualifiers information
-The function `get_inboveg_qualifiers`queries the INBOVEG database for qualifier information on recordings for one or more surveys. 
+The function `get_inboveg_qualifier`queries the INBOVEG database for qualifier information on recordings for one or more surveys.
```
Comment (ThierryO) [inline vignettes/get_data_inboveg.Rmd]:
```suggestion
The function `get_inboveg_qualifier()`queries the INBOVEG database for qualifier information on recordings for one or more surveys. 
```

File: vignettes/get_data_inboveg.Rmd [hunk 8]
```diff
@@ -222,119 +251,132 @@ all_codes <- get_inboveg_classification(con)
...
+Get all layer cover information of all surveys
 
 ```{r, eval = FALSE}
-all_layerinfo <- get_inboveg_layerinfo(con)
+all_layerinfo <- get_inboveg_layer_cover(con)
 ```
 
 ## Layer qualifiers
-The function `get_inboveg_layerqualifier`queries the INBOVEG database for layer qualifier information on recordings  for one or more surveys.
+The function `get_inboveg_layer_qualifier`queries the INBOVEG database for layer qualifier information on recordings  for one or more surveys.
```
Comment (ThierryO) [inline vignettes/get_data_inboveg.Rmd]:
```suggestion
The function `get_inboveg_layer_qualifier()`queries the INBOVEG database for layer qualifier information on recordings  for one or more surveys.
```

File: vignettes/get_data_inboveg.Rmd [hunk 9]
```diff
@@ -222,119 +251,132 @@ all_codes <- get_inboveg_classification(con)
...
 
 ```{r, eval = FALSE}
-alllayerqualifiers <- get_inboveg_layerqualifier(con)
+alllayerqualifiers <- get_inboveg_layer_qualifier(con)
 ```
 
 ## Parent-child relationships between recordings
 
-The function `get_inboveg_relation` queries the INBOVEG database for relational information on recordings for one or more surveys based on parent (classic-chain/bucket) and Child (classic) relationships.
+The function `get_inboveg_relation_recording` queries the INBOVEG database for relational information on recordings for one or more surveys based on parent (classic-chain/bucket) and Child (classic) relationships.
```
Comment (ThierryO) [inline vignettes/get_data_inboveg.Rmd]:
```suggestion
The function `get_inboveg_relation_recording()` queries the INBOVEG database for relational information on recordings for one or more surveys based on parent (classic-chain/bucket) and Child (classic) relationships.
```

**Review by ThierryO — COMMENTED** [review]
Here a a few suggestions

**Comment by ThierryO** [comment]
Note that you can commit suggestions in batch (= multiple suggestions in one commit). You can do that in the "files changed" tab.

**Comment by ElsLommelen** [comment]
> Note that you can commit suggestions in batch (= multiple suggestions in one commit). You can do that in the "files changed" tab.

Too late now, but thanks for the suggestion!

File: R/get_florabank_observations.R [hunk 10]
```diff
@@ -1,19 +1,22 @@
-#' Get all validated observations for one or more taxa from the florabank database
+#' Get all validated observations for one or more taxa from the florabank
+#' database
 #'
 #' This function takes as input a character vector with one or more names of
-#' species either as scientific names and/or Dutch names. By default (fixed = FALSE),
-#' partial matching will be used (the names are prepended with appended with %).
+#' species either as scientific names and/or Dutch names. By default (fixed =
+#' FALSE), partial matching will be used (the names are prepended with appended
```
Comment (hansvancalster) [inline R/get_florabank_observations.R]:
```suggestion
#' FALSE), partial matching will be used (the names are prepended and appended
```

**Review by hansvancalster — APPROVED** [review]
Looks good! Just one question about the renamed `get_inboveg_...` functions. I guess this might break code when users update to the new version of the package. Should we also provide aliases? I think it is OK as it is (i.e. mentioned in NEWS.md), but wanted to check if you had considered this.

**Comment by ElsLommelen** [comment]
@hansvancalster I depreciated these functions, so they can still be used for some time.  After more than one year I will defunct them and keep it in this stage, so users will always be referred to the new names when using the old names.

File: R/connect_inbo_dbase.R (hunk 5 above)
Comment (ElsLommelen) [reply R/connect_inbo_dbase.R]: @ThierryO Finally I changed `<-` back to `=`, as lintr interpreted the argument previewObject as a variable and gave the remark that this variable is defined but not used...

**Comment by florisvdh** [comment]
> Ubuntu 18.04 has reached end of life.

No, this will happen only in April 2023; Ubuntu LTS releases are supported for 5 years. It's only the Ubuntu LTS flavours (like Xubuntu etc) that reach EOL after 3 years.

IMHO it's fine to only test on 20.04 since it's the most used LTS version now.

**Comment by ElsLommelen** [comment]
> IMHO it's fine to only test on 20.04 since it's the most used LTS version now.

This (INBO specific) package has about 3 ubuntu users and they all have 20.04, so I don't care too much here.  Maybe it is more useful to continue this discussion [at checklist](https://github.com/inbo/checklist/pull/76/commits/f53808149962db7178995cca254b8b89958cce8d)?  ;-)

**Pull request merged**